### PR TITLE
Expose set GPU memory fraction per process in CPP API.

### PR DESCRIPTION
**Pull request by vcarpani**
This PR exposes the TORCH.CUDA.SET_PER_PROCESS_MEMORY_FRACTION function that is already exposed in python

cc @ngimel

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

**Comment by vcarpani** [comment]
Hello, any news on merging this?

**Comment by ngimel** [comment]
@vcarpani can you please rebase?

**Comment by ngimel** [comment]
ROCM build error is real, @shintaro-iwasaki what should be the fix?

**Comment by ngimel** [comment]
That's probably from hipification going wrong, so what would be a correct incantation, @shintaro-iwasaki ? Should some functions be added to the hip lists?

**Comment by shintaro-iwasaki** [comment]
Sorry for the late reply. Please let me check it asap.

**Comment by shintaro-iwasaki** [comment]
## 1. Why does it happen?

- `#include <c10/cuda/CUDACachingAllocator.h>` is hipified to `#include <aten/src/ATen/hip/impl/HIPCachingAllocatorMasqueradingAsCUDA.h>` in PyTorch.
- `aten/src/ATen/hip/impl/HIPCachingAllocatorMasqueradingAsCUDA.h` does not implement `setMemoryFraction()`.

    - CUDA side: https://github.com/pytorch/pytorch/blob/master/c10/cuda/CUDACachingAllocator.h
    - AMD side: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/hip/impl/HIPCachingAllocatorMasqueradingAsCUDA.h

## 2. Possible Solutions

### Solution 1a: fix the hipification script

According to this comment, this file should be hipified to use `c10/hip/HIPCachingAllocator.h`, not `HIPCachingAllocatorMasqueradingAsCUDA.h`. We should fix this hipification script. https://github.com/pytorch/pytorch/blob/master/torch/utils/hipify/cuda_to_hip_mappings.py#L7997-L8002

@jeffdaily (or any AMD developers): I don't fully understand the reason behind this conversion, but what behavior is expected? Would you mind teaching us how to fix this issue?

### Solution 1b: directly specify `#include <c10/hip/HIPCachingAllocator.h>`

To unblock this PR (i.e., if we cannot wait for Solution 1a), we can apply a manual hack. The following works as far as I checked.

```cxx
#ifdef USE_ROCM
#include <c10/hip/HIPCachingAllocator.h>
#elif defined(USE_CUDA)
#include <c10/cuda/CUDACachingAllocator.h>
#endif
```

### Solution 2: disable `set_per_process_memory_fraction()` for ROCm

I just checked Solution 1 on AMD GPUs. If Solution 1 does not work for some reasons, the following would be the last resort. Note that `USE_CUDA` is hipified to `USE_ROCM`.

```cxx
#ifdef USE_CUDA
void set_per_process_memory_fraction(float fraction, int64_t device_index) {
#ifdef USE_ROCM
  TORCH_CHECK(false, "set_per_process_memory_fraction() is not supported for ROCm");
#else
  TORCH_CHECK(is_available(), "No CUDA GPUs are available");
  ...
#endif
}
#endif
```

It depends on the functionality of`set_per_process_memory_fraction()`, but perhaps `TORCH_CHECK` can be replaced with a warning version.

**Comment by jeffdaily** [comment]
> I just checked Solution 1 on AMD GPUs. If Solution 1 does not work for some reasons, the following would be the last resort. Note that USE_CUDA is hipified to USE_ROCM.

@shintaro-iwasaki you tried both 1a and 1b?

**Comment by jeffdaily** [comment]
Based on the Note [Masquerading as CUDA]:
https://github.com/pytorch/pytorch/blob/1e3c6f2263c6a679dd11690e0b6c601b795dc759/aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h#L22-L54

The following should be the correct approach.  

```diff
diff --git a/aten/src/ATen/hip/impl/HIPCachingAllocatorMasqueradingAsCUDA.cpp b/aten/src/ATen/hip/impl/HIPCachingAllocatorMasqueradingAsCUDA.cpp
index 46f7d24729..76ee1c1f4b 100644
--- a/aten/src/ATen/hip/impl/HIPCachingAllocatorMasqueradingAsCUDA.cpp
+++ b/aten/src/ATen/hip/impl/HIPCachingAllocatorMasqueradingAsCUDA.cpp
@@ -13,5 +13,9 @@ void recordStreamMasqueradingAsCUDA(const DataPtr& ptr, HIPStreamMasqueradingAsC
   HIPCachingAllocator::recordStream(ptr, stream.hip_stream());
 }

+void setMemoryFractionMasqueradingAsCUDA(double fraction, int device) {
+  HIPCachingAllocator::setMemoryFraction(fraction, device);
+}
+
 } // namespace HIPCachingAllocatorMasqueradingAsCUDA
 }} // namespace c10::hip
diff --git a/aten/src/ATen/hip/impl/HIPCachingAllocatorMasqueradingAsCUDA.h b/aten/src/ATen/hip/impl/HIPCachingAllocatorMasqueradingAsCUDA.h
index 3aaa9d06c5..9a108fc258 100644
--- a/aten/src/ATen/hip/impl/HIPCachingAllocatorMasqueradingAsCUDA.h
+++ b/aten/src/ATen/hip/impl/HIPCachingAllocatorMasqueradingAsCUDA.h
@@ -12,6 +12,7 @@ namespace HIPCachingAllocatorMasqueradingAsCUDA {

 C10_HIP_API Allocator* get();
 C10_HIP_API void recordStreamMasqueradingAsCUDA(const DataPtr& ptr, HIPStreamMasqueradingAsCUDA stream);
+C10_HIP_API void setMemoryFractionMasqueradingAsCUDA(double fraction, int device);

 } // namespace HIPCachingAllocatorMasqueradingAsCUDA
 } // namespace hip
diff --git a/torch/utils/hipify/cuda_to_hip_mappings.py b/torch/utils/hipify/cuda_to_hip_mappings.py
index 6f4c935047..e2c657ce22 100644
--- a/torch/utils/hipify/cuda_to_hip_mappings.py
+++ b/torch/utils/hipify/cuda_to_hip_mappings.py
@@ -7994,8 +7994,6 @@ PYTORCH_SPECIFIC_MAPPINGS = collections.OrderedDict(
             "CUDAMultiStreamGuard",
             ("HIPMultiStreamGuardMasqueradingAsCUDA", API_PYTORCH),
         ),
-        # Only get needs to be transformed this way; all the other ones can go
-        # straight to the normal versions hip::HIPCachingAllocator
         (
             "cuda::CUDACachingAllocator::get",
             ("hip::HIPCachingAllocatorMasqueradingAsCUDA::get", API_PYTORCH),
@@ -8018,6 +8016,20 @@ PYTORCH_SPECIFIC_MAPPINGS = collections.OrderedDict(
                 API_PYTORCH,
             ),
         ),
+        (
+            "cuda::CUDACachingAllocator::setMemoryFraction",
+            (
+                "hip::HIPCachingAllocatorMasqueradingAsCUDA::setMemoryFractionMasqueradingAsCUDA",
+                API_PYTORCH,
+            ),
+        ),
+        (
+            "CUDACachingAllocator::setMemoryFraction",
+            (
+                "HIPCachingAllocatorMasqueradingAsCUDA::setMemoryFractionMasqueradingAsCUDA",
+                API_PYTORCH,
+            ),
+        ),
         ("cuda::CUDAStream", ("hip::HIPStreamMasqueradingAsCUDA", API_PYTORCH)),
         ("CUDAStream", ("HIPStreamMasqueradingAsCUDA", API_PYTORCH)),
         (
```

**Comment by jeffdaily** [comment]
I've tried solution 1a and 1b, but both fail with undefined references at link time.

1a -- where I update the hipify mappings and implement a masquerade pass-through for setMemoryFractionMasqueradingAsCUDA

```../lib/libtorch_cpu.so: undefined reference to `c10::hip::HIPCachingAllocatorMasqueradingAsCUDA::setMemoryFractionMasqueradingAsCUDA(double, int)'```

1b -- where I directly specify `#include <c10/hip/HIPCachingAllocator.h>` in torch/csrc/api/src/cuda.cpp, I get 

```../lib/libtorch_cpu.so: undefined reference to `c10::hip::HIPCachingAllocator::setMemoryFraction(double, int)'```

**Comment by jeffdaily** [comment]
Shouldn't this be exposed through CUDAHooks like the rest of the APIs?

**Comment by shintaro-iwasaki** [comment]
@jeffdaily: That's right. I checked it locally and it 1b did not work. Thanks for your check.

**Comment by ngimel** [comment]
Yeah we can expose it through CUDAHooks (together with other functions in HIPCachingAllocatorMasqueradingAsCUDA.h currently), but let's not block this PR on this?

**Comment by jeffdaily** [comment]
> Yeah we can expose it through CUDAHooks (together with other functions in HIPCachingAllocatorMasqueradingAsCUDA.h currently), but let's not block this PR on this?

Since solution 1a and 1b didn't work, the only alternative seems to be (2) where we ifdef away the rocm ability to make this same call.  If CUDAHooks is the correct implementation, why merge this currently less-than-ideal PR?

**Comment by jeffdaily** [comment]
Does this PR also need a unit test for the newly exposed API?

**Comment by jeffdaily** [comment]
I don't have permission to push to this PR branch, but I propose the following diff that can be applied as the next commit in this PR.  I have confirmed that this will compile for ROCm.

```diff
diff --git a/aten/src/ATen/cuda/detail/CUDAHooks.cpp b/aten/src/ATen/cuda/detail/CUDAHooks.cpp
index 93a23ec6a7..238c692493 100644
--- a/aten/src/ATen/cuda/detail/CUDAHooks.cpp
+++ b/aten/src/ATen/cuda/detail/CUDAHooks.cpp
@@ -408,6 +408,10 @@ void CUDAHooks::deviceSynchronize(int64_t device_index) const {
   c10::cuda::device_synchronize();
 }

+void CUDAHooks::setMemoryFraction(double fraction, int64_t device_index) const {
+  c10::cuda::CUDACachingAllocator::setMemoryFraction(fraction, int(device_index));
+}
+
 // Sigh, the registry doesn't support namespaces :(
 using at::CUDAHooksRegistry;
 using at::RegistererCUDAHooksRegistry;
diff --git a/aten/src/ATen/cuda/detail/CUDAHooks.h b/aten/src/ATen/cuda/detail/CUDAHooks.h
index 5aa2721170..42d533b938 100644
--- a/aten/src/ATen/cuda/detail/CUDAHooks.h
+++ b/aten/src/ATen/cuda/detail/CUDAHooks.h
@@ -50,6 +50,7 @@ struct CUDAHooks : public at::CUDAHooksInterface {
   void cuFFTClearPlanCache(int64_t device_index) const override;
   int getNumGPUs() const override;
   void deviceSynchronize(int64_t device_index) const override;
+  void setMemoryFraction(double fraction, int64_t device_index) const override;
 };

 }}} // at::cuda::detail
diff --git a/aten/src/ATen/detail/CUDAHooksInterface.h b/aten/src/ATen/detail/CUDAHooksInterface.h
index 1303b9f8c8..6e1aeb2d97 100644
--- a/aten/src/ATen/detail/CUDAHooksInterface.h
+++ b/aten/src/ATen/detail/CUDAHooksInterface.h
@@ -187,6 +187,10 @@ struct TORCH_API CUDAHooksInterface {
   virtual void deviceSynchronize(int64_t /*device_index*/) const {
     TORCH_CHECK(false, "Cannot synchronize CUDA device without ATen_cuda library. ", CUDA_HELP);
   }
+
+  virtual void setMemoryFraction(double /*fraction*/, int64_t /*device_index*/) const {
+    TORCH_CHECK(false, "Cannot set CUDA device memory fraction limit without ATen_cuda library. ", CUDA_HELP);
+  }
 };

 // NB: dummy argument to suppress "ISO C++11 requires at least one argument
diff --git a/torch/csrc/api/include/torch/cuda.h b/torch/csrc/api/include/torch/cuda.h
index 30836a9372..54eef0518d 100644
--- a/torch/csrc/api/include/torch/cuda.h
+++ b/torch/csrc/api/include/torch/cuda.h
@@ -26,11 +26,8 @@ void TORCH_API manual_seed_all(uint64_t seed);
 /// Waits for all kernels in all streams on a CUDA device to complete.
 void TORCH_API synchronize(int64_t device_index = -1);

-#ifdef USE_CUDA
 /// Sets memory fraction for a process.
-void TORCH_API
-set_per_process_memory_fraction(float fraction, int64_t device_index = -1);
-#endif
+void TORCH_API set_per_process_memory_fraction(float fraction, int64_t device_index = -1);

 } // namespace cuda
 } // namespace torch
diff --git a/torch/csrc/api/src/cuda.cpp b/torch/csrc/api/src/cuda.cpp
index 38b24ea487..a5230f9235 100644
--- a/torch/csrc/api/src/cuda.cpp
+++ b/torch/csrc/api/src/cuda.cpp
@@ -4,10 +4,6 @@
 #include <c10/core/DeviceGuard.h>
 #include <c10/util/irange.h>

-#ifdef USE_CUDA
-#include <c10/cuda/CUDACachingAllocator.h>
-#endif
-
 #include <cstddef>

 namespace torch {
@@ -65,7 +61,6 @@ void synchronize(int64_t device_index) {
   at::detail::getCUDAHooks().deviceSynchronize(device_index);
 }

-#ifdef USE_CUDA
 void set_per_process_memory_fraction(float fraction, int64_t device_index) {
   TORCH_CHECK(is_available(), "No CUDA GPUs are available");
   int64_t num_gpus = cuda::device_count();
@@ -77,10 +72,8 @@ void set_per_process_memory_fraction(float fraction, int64_t device_index) {
       fraction >= 0.0 || fraction <= 1.0,
       "Memory fraction should be between 0 and 1, got: ",
       fraction);
-  c10::cuda::CUDACachingAllocator::setMemoryFraction(
-      fraction, int(device_index));
+  at::detail::getCUDAHooks().setMemoryFraction(fraction, device_index);
 }
-#endif

 } // namespace cuda
 } // namespace torch
```

**Comment by ngimel** [comment]
Thanks so much @jeffdaily, this is incredibly helpful! @vcarpani can you please apply this patch?

**Comment by vcarpani** [comment]
> Thanks so much @jeffdaily, this is incredibly helpful! @vcarpani can you please apply this patch?

Hello, I applied the changes, could someone approve the CI job?

**Comment by jeffdaily** [comment]
@vcarpani please rebase (use command `@pytorchbot rebase`).  This is necessary to fix a different rocm ci failure due to a  missing docker image.

**Pull request closed without merging**
